diesel_derives/src/sql_type.rs

```diff
@@ -0,0 +1,133 @@
+use quote::Tokens;
+use syn;
+
+use util::*;
+
+pub fn derive(item: syn::DeriveInput) -> Tokens {
+    let struct_ty = struct_ty(item.ident.clone(), &item.generics);
+    let item_name = item.ident.as_ref().to_uppercase();
+    let generics = &item.generics;
+    let pg_tokens = pg_tokens(&item.attrs, generics, &struct_ty);
+    let sqlite_tokens = sqlite_tokens(&item.attrs, generics, &struct_ty);
+    let mysql_tokens = mysql_tokens(&item.attrs, generics, &struct_ty);
+
+    wrap_item_in_const(
+        format!("_IMPL_SQL_TYPE_FOR_{}", item_name).into(),
+        quote!(
+            impl #generics diesel::types::NotNull for #struct_ty {}
+            impl #generics diesel::types::SingleValue for #struct_ty {}
+            #pg_tokens
+            #sqlite_tokens
+            #mysql_tokens
+        ),
+    )
+}
+
+fn pg_tokens(
+    attrs: &[syn::Attribute],
+    generics: &syn::Generics,
+    struct_ty: &syn::Ty,
+) -> Option<Tokens> {
+    use syn::{Lit, MetaItem, NestedMetaItem};
+
+    fn error() -> ! {
+        panic!(
+            "#[postgres] must be in the form \
+             #[postgres(oid = \"1\", array_oid = \"2\")] \
+             or #[postgres(type_name = \"my_type\")]"
+        );
+    }
+
+    if cfg!(not(feature = "postgres")) {
```

Comment:
I case that "postgres" is not enabled the `#[postgres]` attribute is not checked. I think we should either check the attribute unconditionally or report an error if the attribute exists but "postgres" is not enabled.

Reply:
That would mean Diesel would error every time you compile without all three backends. You should be able to declare how a backend supports a type, and then only have the code generated if that backend is enabled for Diesel.

Reply:
Then we should at least check if the attribute would be valid, without generating the actual impl.